File: frontend/lib/mappers/net-worth-mapper.ts
/**
 * Net Worth Mapper - Stage 4 Net Worth Intelligence Workspace
 *
 * Transforms backend DTOs to ViewModels for the Net Worth Intelligence Workspace.
 * This is the ONLY location where DTO to ViewModel mapping occurs.
 *
 * Architecture Flow: Backend → API → DTO → Mapper → ViewModel
 */

import type {
  NetWorthViewModel,
  NetWorthCompositionViewModel,
  NetWorthBreakdownItemViewModel,
  NetWorthHistoricalSnapshotViewModel,
  NetWorthTrendViewModel,
  NetWorthInsightViewModel,
  NetWorthEvidenceChainViewModel,
  NetWorthFiltersViewModel,
  NetWorthNavigationViewModel,
} from '../../types/net-worth-view-model';

// ===== DTO Types (from backend) =====

interface NetWorthBreakdownItemDTO {
  id: string;
  name: string;
  type: string;
  balance_paise: number;
  percentage: number;
  contribution_paise: number;
}

interface NetWorthCompositionDTO {
  total_assets_paise: number;
  total_liabilities_paise: number;
  asset_breakdown: NetWorthBreakdownItemDTO[];
  liability_breakdown: NetWorthBreakdownItemDTO[];
}

interface NetWorthHistoricalSnapshotDTO {
  date: string;
  net_worth_paise: number;
  assets_paise: number;
  liabilities_paise: number;
}

type NetWorthTrendDirection = 'up' | 'down' | 'flat';

interface NetWorthTrendDTO {
  direction: NetWorthTrendDirection;
  percentage_change: number;
  period: string;
}

type NetWorthInsightType = 'positive' | 'warning' | 'info' | 'alert';
type NetWorthInsightSeverity = 'low' | 'medium' | 'high';

interface NetWorthInsightDTO {
  type: NetWorthInsightType;
  severity: NetWorthInsightSeverity;
  message: string;
  action_url?: string;
}

interface NetWorthEvidenceItemDTO {
  type: string;
  summary: string;
  source: string;
  confidence?: number;
}

interface NetWorthCalculationStepDTO {
  name: string;
  description: string;
  inputs: Record<string, unknown>;
  outputs: Record<string, unknown>;
}

interface NetWorthEvidenceChainDTO {
  summary: string;
  evidence: NetWorthEvidenceItemDTO[];
  calculation_steps: NetWorthCalculationStepDTO[];
  source_references: string[];
  confidence_score: number;
}

interface NetWorthDTO {
  total_net_worth_paise: number;
  total_assets_paise: number;
  total_liabilities_paise: number;
  composition: NetWorthCompositionDTO;
  trend?: NetWorthTrendDTO;
  insights: NetWorthInsightDTO[];
  evidence_chain?: NetWorthEvidenceChainDTO;
}

/**
 * Net Worth Mapper Interface
 * Defines the contract for DTO to ViewModel mapping
 */
export interface INetWorthMapper {
  /**
   * Map a single Net Worth DTO to ViewModel
   */
  mapNetWorthDTO(dto: NetWorthDTO): NetWorthViewModel;

  /**
   * Map composition DTO to ViewModel
   */
  mapCompositionDTO(dto: NetWorthCompositionDTO): NetWorthCompositionViewModel;

  /**
   * Map historical snapshots DTOs to ViewModels
   */
  mapHistoricalSnapshots(
    dtos: NetWorthHistoricalSnapshotDTO[]
  ): NetWorthHistoricalSnapshotViewModel[];

  /**
   * Map trend DTO to ViewModel
   */
  mapTrendDTO(dto: NetWorthTrendDTO | null | undefined): NetWorthTrendViewModel | undefined;

  /**
   * Map insights DTOs to ViewModels
   */
  mapInsights(dtos: NetWorthInsightDTO[]): NetWorthInsightViewModel[];

  /**
   * Map evidence chain DTO to ViewModel
   */
  mapEvidenceChain(
    dto: NetWorthEvidenceChainDTO | null | undefined
  ): NetWorthEvidenceChainViewModel | undefined;
}

/**
 * Net Worth Mapper Implementation
 * Transforms backend net worth data to ViewModels
 */
export class NetWorthMapper implements INetWorthMapper {
  /**
   * Map a single Net Worth DTO to ViewModel
   * @param dto - Net Worth data from API
   * @returns NetWorthViewModel for presentation
   */
  mapNetWorthDTO(dto: NetWorthDTO): NetWorthViewModel {
    return {
      total_net_worth_paise: dto.total_net_worth_paise,
      total_assets_paise: dto.total_assets_paise,
      total_liabilities_paise: dto.total_liabilities_paise,
      composition: this.mapCompositionDTO(dto.composition),
      trend: this.mapTrendDTO(dto.trend),
      insights: this.mapInsights(dto.insights),
      evidence_chain: this.mapEvidenceChain(dto.evidence_chain),
      filters: this.createDefaultFilters(),
      navigation: this.createDefaultNavigation(),
    };
  }

  /**
   * Map composition DTO to ViewModel
   */
  mapCompositionDTO(dto: NetWorthCompositionDTO): NetWorthCompositionViewModel {
    return {
      total_assets_paise: dto.total_assets_paise,
      total_liabilities_paise: dto.total_liabilities_paise,
      asset_breakdown: dto.asset_breakdown.map((item) => this.mapBreakdownItem(item)),
      liability_breakdown: dto.liability_breakdown.map((item) =>
        this.mapBreakdownItem(item)
      ),
    };
  }

  /**
   * Map a single breakdown item DTO to ViewModel
   */
  private mapBreakdownItem(dto: NetWorthBreakdownItemDTO): NetWorthBreakdownItemViewModel {
    return {
      id: dto.id,
      name: dto.name,
      type: dto.type,
      balance_paise: dto.balance_paise,
      percentage: dto.percentage,
      contribution_paise: dto.contribution_paise,
    };
  }

  /**
   * Map historical snapshots DTOs to ViewModels
   */
  mapHistoricalSnapshots(
    dtos: NetWorthHistoricalSnapshotDTO[]
  ): NetWorthHistoricalSnapshotViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      date: dto.date,
      net_worth_paise: dto.net_worth_paise,
      assets_paise: dto.assets_paise,
      liabilities_paise: dto.liabilities_paise,
    }));
  }

  /**
   * Map trend DTO to ViewModel
   */
  mapTrendDTO(
    dto: NetWorthTrendDTO | null | undefined
  ): NetWorthTrendViewModel | undefined {
    if (!dto) {
      return undefined;
    }
    return {
      direction: dto.direction,
      percentage_change: dto.percentage_change,
      period: dto.period,
    };
  }

  /**
   * Map insights DTOs to ViewModels
   */
  mapInsights(dtos: NetWorthInsightDTO[]): NetWorthInsightViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      type: dto.type,
      severity: dto.severity,
      message: dto.message,
      action_url: dto.action_url,
    }));
  }

  /**
   * Map evidence chain DTO to ViewModel
   */
  mapEvidenceChain(
    dto: NetWorthEvidenceChainDTO | null | undefined
  ): NetWorthEvidenceChainViewModel | undefined {
    if (!dto) {
      return undefined;
    }
    return {
      summary: dto.summary,
      evidence: dto.evidence.map((item) => ({
        type: item.type,
        summary: item.summary,
        source: item.source,
        confidence: item.confidence,
      })),
      calculation_steps: dto.calculation_steps.map((step) => ({
        name: step.name,
        description: step.description,
        inputs: step.inputs,
        outputs: step.outputs,
      })),
      source_references: dto.source_references,
      confidence_score: dto.confidence_score,
    };
  }

  /**
   * Create default filters
   */
  private createDefaultFilters(): NetWorthFiltersViewModel {
    return {
      date_range: undefined,
      account_types: undefined,
      period: undefined,
    };
  }

  /**
   * Create default navigation
   */
  private createDefaultNavigation(): NetWorthNavigationViewModel {
    return {
      deep_link: '/net-worth',
      cross_references: {
        accounts: '/accounts',
        investments: '/investments',
        loans: '/loans',
        credit_cards: '/credit-cards',
      },
    };
  }
}

// Export singleton instance
export const netWorthMapper = new NetWorthMapper();

File: backend/src/core/mappers/networth_mapper.py
"""Net Worth Mapper

Transforms net worth domain objects into NetWorthDTO instances.
This is the ONLY location where net worth API responses are constructed.
"""

from typing import Any

from src.core.dtos.net_worth_dto import (
    NetWorthBreakdownItemDTO,
    NetWorthCompositionDTO,
    NetWorthDTO,
    NetWorthHistoricalSnapshotDTO,
    NetWorthInsightDTO,
    NetWorthTrendDTO,
)


class NetWorthMapper:
    """Mapper for net worth domain objects to DTOs."""

    @staticmethod
    def to_dto(net_worth_data: dict[str, Any]) -> NetWorthDTO:
        """Convert net worth data to NetWorthDTO."""
        return NetWorthDTO(
            total_net_worth_paise=net_worth_data.get("total_net_worth_paise", 0),
            total_assets_paise=net_worth_data.get("total_assets_paise", 0),
            total_liabilities_paise=net_worth_data.get("total_liabilities_paise", 0),
            composition=NetWorthMapper._to_composition_dto(
                net_worth_data.get("composition", {})
            ),
            trend=NetWorthMapper._to_trend_dto(net_worth_data.get("trend")),
            insights=NetWorthMapper._to_insights(net_worth_data.get("insights", [])),
            evidence_chain=net_worth_data.get("evidence_chain"),
        )

    @staticmethod
    def _to_composition_dto(composition_data: dict[str, Any]) -> NetWorthCompositionDTO:
        """Convert composition data to NetWorthCompositionDTO."""
        return NetWorthCompositionDTO(
            total_assets_paise=composition_data.get("total_assets_paise", 0),
            total_liabilities_paise=composition_data.get("total_liabilities_paise", 0),
            asset_breakdown=NetWorthMapper._to_breakdown_items(
                composition_data.get("asset_breakdown", [])
            ),
            liability_breakdown=NetWorthMapper._to_breakdown_items(
                composition_data.get("liability_breakdown", [])
            ),
        )

    @staticmethod
    def _to_breakdown_items(
        items: list[dict[str, Any]],
    ) -> list[NetWorthBreakdownItemDTO]:
        """Convert breakdown items to NetWorthBreakdownItemDTO list."""
        return [
            NetWorthBreakdownItemDTO(
                id=item.get("id", ""),
                name=item.get("name", "Unknown"),
                type=item.get("type", "other"),
                balance_paise=item.get("balance_paise", 0),
                percentage=item.get("percentage", 0.0),
                contribution_paise=item.get("contribution_paise", 0),
            )
            for item in items
        ]

    @staticmethod
    def _to_trend_dto(trend_data: dict[str, Any] | None) -> NetWorthTrendDTO | None:
        """Convert trend data to NetWorthTrendDTO."""
        if not trend_data:
            return None
        return NetWorthTrendDTO(
            direction=trend_data.get("direction", "flat"),
            percentage_change=trend_data.get("percentage_change", 0.0),
            period=trend_data.get("period", "1M"),
        )

    @staticmethod
    def _to_insights(insights: list[dict[str, Any]]) -> list[NetWorthInsightDTO]:
        """Convert insights to NetWorthInsightDTO list."""
        return [
            NetWorthInsightDTO(
                type=insight.get("type", "info"),
                severity=insight.get("severity", "medium"),
                message=insight.get("message", ""),
                action_url=insight.get("action_url"),
            )
            for insight in insights
        ]

    @staticmethod
    def to_historical_snapshot_dto(
        snapshot_data: dict[str, Any],
    ) -> NetWorthHistoricalSnapshotDTO:
        """Convert historical snapshot data to NetWorthHistoricalSnapshotDTO."""
        return NetWorthHistoricalSnapshotDTO(
            date=snapshot_data.get("date", ""),
            net_worth_paise=snapshot_data.get("net_worth_paise", 0),
            assets_paise=snapshot_data.get("assets_paise", 0),
            liabilities_paise=snapshot_data.get("liabilities_paise", 0),
        )
